torch/_decomp/__init__.py

```diff
@@ -170,6 +177,33 @@ def get_decompositions(
     return decompositions
 
 
+def enable_pre_autograd_decomposition(
+    aten_ops: Sequence[Union[torch._ops.OpOverload, torch._ops.OpOverloadPacket]]
+) -> Dict[torch._ops.OpOverload, Callable]:
+    """
+    Register a decomposition for an operator to be used in the pre-autograd
+    decomposition pass. Its backward implementation will be generated according
+    to the decomposed forward graph.
+
+    NB: The aten ops passed to this function must be CompositeExplicitAutograd ops
+    """
+
+    decomps = get_decompositions(aten_ops)
+    assert len(decomps) == len(
+        aten_ops
+    ), "Not all operators have decompositions registered"
+
+    for opo, fn in decomps.items():
+        # For CompositeExplicitAutograd ops, we must register the decomposition function under 'Autograd' key
+        # There is some magic here, see resolve_key() in torch/_ops.py for more details
+        opo.py_impl(torch._C.DispatchKey.Autograd)(fn)
```

Comment:
So, this is a bit goofy; once you register it in this way, it is ALWAYS turned on when you run python dispatcher. So it's not like the decomposition table, where you can transparently change it out depending on what tracing context you're in.

To start, I'd prefer it if we only maintain a global set of pre-autograd decompositions that is not configurable. Making this configurable is only slightly more involved, as we must flush the python dispatcher cache to ensure we recompute all the dispatch keys when we change the registrations.

Reply:
Based on chat discussion, it seems like people greatly prefer it if it is configurable. In that case, let's just implement the cache flushing mechanism.